src/Guard/TelegramAuthenticator.php

```diff
@@ -0,0 +1,141 @@
+<?php
+
+namespace BoShurik\TelegramBotBundle\Guard;
+
+use BoShurik\TelegramBotBundle\Guard\UserProviderInterface as BoShurikUserProviderInterface;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\Routing\Generator\UrlGeneratorInterface;
+use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
+use Symfony\Component\Security\Core\Exception\CustomUserMessageAuthenticationException;
+use Symfony\Component\Security\Core\Security;
+use Symfony\Component\Security\Core\User\UserInterface;
+use Symfony\Component\Security\Core\User\UserProviderInterface;
+use Symfony\Component\Security\Guard\Authenticator\AbstractFormLoginAuthenticator;
+use Symfony\Component\Security\Http\Util\TargetPathTrait;
+
+class TelegramAuthenticator extends AbstractFormLoginAuthenticator
+{
+    use TargetPathTrait;
+
+    private const REQUIRED_FIELDS = [
+        'id',
+        'first_name',
+        'last_name',
+        'auth_date',
+        'hash',
+    ];
+
+    /**
+     * @var string
+     */
+    private $secret;
+
+    /**
+     * @var BoShurikUserProviderInterface
+     */
+    private $userProvider;
+
+    /**
+     * @var UrlGeneratorInterface
+     */
+    private $urlGenerator;
+
+    public function __construct(BoShurikUserProviderInterface $userProvider, UrlGeneratorInterface $urlGenerator, string $telegramBotToken)
+    {
+        $this->userProvider = $userProvider;
+        $this->urlGenerator = $urlGenerator;
+        $this->secret = hash('sha256', $telegramBotToken, true);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function supports(Request $request)
+    {
+        $route = $request->attributes->get('_route');
+        $data = $request->query->all();
+
+        return $route === '_telegram_login'
+            && array_intersect(self::REQUIRED_FIELDS, array_keys($data)) === self::REQUIRED_FIELDS;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    final public function getCredentials(Request $request)
+    {
+        $credentials = $request->query->all();
+
+        $request->getSession()->set(
+            Security::LAST_USERNAME,
+            $credentials['username'] ?? null
+        );
```

Comment:
I think it's unnecessary as this value it not used and user may use different username at the site and at telegram